modules/features2d/src/sift.dispatch.cpp

```diff
@@ -522,6 +503,30 @@ void SIFT_Impl::detectAndCompute(InputArray _image, InputArray _mask,
     {
         // filter keypoints by mask
         //KeyPointsFilter::runByPixelsMask( keypoints, mask );
+
+        // assign suitable octaves
+        for (size_t i = 0; i < keypoints.size(); i++ )
+        {
+            int octv, layer;
+            KeyPoint& kpt = keypoints[i];
+            float octv_layer = std::log(kpt.size / sigma) / std::log(2.) - 1;
```

Comment:
The idea of octave value reevaluation is definitely good. However it would be better to retain the descriptor evaluation as is(to allow the user evaluation of non-standard keypoints if one wants to filter and/or modify the keypoint set in some way).
I suggest moving the octave reevaluation to the implementation of the test(the tough part is that octave update should be done for SIFT test instantiation only cause it looks like AKAZE have different meaning for octave value).

Reply:
I can change the execution path by checking `getDefaultName()` (by the way it seems that SIFT doesn't override this function, so I will define it) but isn't it awkward to write such if-statements in `TEST_P(DetectorScaleInvariance, scale)`?

Reply:
`DetectorScaleInvariance` typedef could be replaced with class public inheritance with additional method that will retain all the code of the test implementation and will have a "pointer to a function" parameter defaulted to existing `scaleKeyPoints`. So the existing test could be defined as simple call to that method and SIFT test(implemented as a separate `TEST_P`) could be provided with a specific scaling function.